pattern/src/lib.rs

```diff
@@ -1,9 +1,107 @@
+//! This crate implements a macro-based DSL for matching netlists.
+//!
+//! The DSL features programmable patterns, which enables many features that
+//! wouldn't otherwise be possible. For example:
+//! - [`PZExt`] can match zero extensions, even though they don't correspond
+//!   to a cell at all, and are achieved by appropriately mixing [`Net`]s
+//!   within a [`Value`]
+//! - patterns which correspond to bitwise cells, such as [`PAnd`], can match
+//!   a [`Value`] even if it is made up from [`Net`]s coming from many distinct
+//!   [`And`] cells.
+//! - specialized patterns can be easily implemented, such as [`PPow2@a`],
+//!   which matches a constant power of 2 and captures the exponent into `a`.
```

Comment:
This section isn't technically but I feel like it's motivated in a confusing way: I don't think "programmable pattern" is a widely known term of art, and I feel like it's written as a comparison to something else that the reader may not already have in mind as a working assumption of how it works. (I suppose it was ML-style pattern matching?)

In Amaranth I've avoided writing documentation like this because reading a *diff* to something is fundamentally harder than reading *the thing it's diffing towards*.

Reply:
I am comparing to Rust's `match` statement, which I assume the reader is familiar with.

Reply:
I still feel that describing the behavior as it is would be less cognitive load than describing the behavior as a difference to an existing concept.